keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,146 @@
+  // The name is "<resource>.<group>". The API validation will enforce it by
+  // comapring the name with the spec.resource.
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+type StorageStateSpec {
+  // The resource this StoageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of versions that the persisted objects might be encoded in.
+  // If empty, it means the stored versions are unknown.
+  // +optional
+  StoredVersionsHashes []string
+  // The hash value of the storage version, as shown in the discovery document
+  // served by the API server.
+  // Storage Version is the version to which objects are converted to
+  // before persisted.
+  StorageVersionHash string
+}
+```
+
+We had considered making `StoredVersionsHashes` part of the `spec` instead of
+the `status`, because the stored versions cannot be deduced immediately by
+inspecting Kubernetes API. However, we make it part of the status because
+* Its value can be determined by inspecting the `StorageVersionMigration` API.
+  It just takes longer, i.e., until a migration has completed. 
+* Leaving it in the status is consistent with the [CRD.status.StoredVersions][]
+  API.
+
+[CRD.status.StoredVersions]:https://github.com/kubernetes/kubernetes/blob/697c2316faaabae8ef8371032b60be65d7795e68/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L305
+  
+## Storage migration triggering controller
+
+We will add a controller that monitors the discovery documents and the
+storageStates to trigger storage version migration.
+
+The controller fetches the discovery document periodically. For each resource,
+it compares the storageVersion in the discovery document with that recorded in
+the storageState. If they differ, or if the storageState does not exist, the
+controller
+  * stops any ongoing migration of this resource by deleting existing
+    storageVersionMigrations,
+  * creates a new storageVersionMigration for this resource,
+  * in a single operation, updates (or creates) the
+    storageState.status.StorageVersionHash to the value shown in the discovery
+    document, and adds the hash to the storageState.status.storedVersionHashes.
+
+The refresh rate of the discovery document needs to satisfy two conflicting
+goals: *i)* not causing too much traffic to apiservers, and *ii)* not missing
+storage version configuration changes. 10 minutes is a good middle ground.
+
+The controller uses an informer to monitor all storageVersionMigrations. If a
+migration completes, the controller finds the storageState of that resource,
+updates its status.storedVersionHashes to equal its status.storageVersionHash.
+
+If the controller crashloops and misses storage version changes, the
+storageState.status.storedVersionHashes can be stale. Thus, the controller
+should recreate storageState objects with empty storedVersionHashes when it
+bootstraps to avoid acting on stale information.
```

Comment:
so it would lose the state / history, that could cause missed auto-triggering, no?

Reply:
"empty storedVersionHashes" will trigger a mass migration to restore state.

Reply:
i might not fully understand the intended lifecycle of the storageState object, when it is created, when it is updated, when its field storedVersionHashes is cleared, when it is deleted. better to make it more explicit.

one quick question I have is, when the controller bootstraps (starts), does it "recreate storageState objects with empty storedVersionHashes"? it seems to be intended as described in the prior paragraph. however, that means the object would only live within the controller's lifespan, that seems to be very fragile.